include/fiction/algorithms/simulation/sidb/maximal_defect_influence_distance.hpp

```diff
@@ -0,0 +1,260 @@
+    lyt.foreach_cell([&cells](const auto& cell) { cells.push_back(cell); });
+    std::sort(cells.begin(), cells.end());
+
+    const quickexact_params<Lyt> params{sim_params.physical_params, automatic_base_number_detection::OFF};
+    auto                         simulation_results = quickexact(lyt, params);
+
+    const auto min_energy          = minimum_energy(simulation_results.charge_distributions);
+    uint64_t   charge_index_layout = 0;
+
+    for (auto& lyt_result : simulation_results.charge_distributions)
+    {
+        if (round_to_n_decimal_places(lyt_result.get_system_energy(), 6) == round_to_n_decimal_places(min_energy, 6))
+        {
+            lyt_result.charge_distribution_to_index_general();
+            charge_index_layout = lyt_result.get_charge_index_and_base().first;
+        }
+    }
+
+    const auto [nw, se] = bounding_box_siqad(layout);
+
+    auto north_west = coordinate<Lyt>{};
+    auto south_east = coordinate<Lyt>{};
+
+    north_west.x = nw.x - sim_params.additional_scanning_area.x;
+    north_west.y = nw.y - sim_params.additional_scanning_area.y;
+
+    south_east.x = se.x + sim_params.additional_scanning_area.x;
+    south_east.y = se.y + sim_params.additional_scanning_area.y;
+
+    auto                            defect_cell = north_west;
+    std::vector<typename Lyt::cell> defect_cells{};
+
+    while (defect_cell <= south_east)
+    {
+        if (lyt.get_cell_type(defect_cell) == sidb_technology::cell_type::EMPTY)
+        {
+            defect_cells.push_back(defect_cell);
+        }
+
+        if (defect_cell.x < south_east.x)
+        {
+            defect_cell.x += 1;
+        }
+        else if ((defect_cell.x == south_east.x) && defect_cell.z == 0)
+        {
+            defect_cell.z += 1;
+            defect_cell.x = north_west.x;
+        }
+
+        else if ((defect_cell.x == south_east.x) && defect_cell.z == 1)
+        {
+            defect_cell.x = north_west.x;
+            defect_cell.y += 1;
+            defect_cell.z = 0;
+        }
+    }
+
+    // If the number of threads is initially set to zero, the simulation is run with one thread.
+    const uint64_t num_threads = std::max(sim_params.number_threads, uint64_t{1});
+
+    std::vector<std::thread> threads{};
+    threads.reserve(num_threads);
+    std::mutex mutex{};  // used to control access to shared resources
+
+    const auto number_per_thread = (defect_cells.size() - (defect_cells.size() % num_threads)) / num_threads;
+    const auto number_last       = defect_cells.size() % num_threads;
+
+    for (uint64_t z = 0u; z < num_threads; ++z)
+    {
+        threads.emplace_back(
+            [&, z, defect_cells]
+            {
+                for (auto i = z * number_per_thread; i < (z + 1) * number_per_thread; i++)
+                {
+                    const auto defect = defect_cells[i];
+
+                    sidb_defect_layout lyt_defect{};
+
+                    if (layout.num_cells() != 0)
```

Comment:
I don't think you need this check. If the number of cells is 0, `foreach_cell` will simply not run.